lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,484 @@
+
+        cr:rectangle( bw/2, bw/2, max_x, max_y )
+
+        for i=1, #max_row do
+            cr:move_to(bw/2, math.max(0, sh - spacing/2))
+            cr:line_to(max_x, math.max(0, sh - spacing/2))
+
+            sh = sh + max_row[i] + spacing
+        end
+
+        for i=1, #max_col do
+            cr:move_to(math.max(0, sw - spacing/2), bw/2)
+            cr:line_to(math.max(0, sw - spacing/2), max_y)
+
+            sw = sw + max_col[i] + spacing
+        end
+
+        cr:stroke()
+    end
+end
+
+local function get_next_empty(self, row, column)
+    row, column = row or 1, column or 1
+    local cc = self._private.column_count
+    for i = row, math.huge do
+        local r = self._private.rows[i]
+
+        if not r then
+            r = {}
+            self._private.rows[i] = r
+            self._private.row_count = i
+
+            return i, 1
+        end
+
+        for j = column, cc do
+            if not r[j] then
+                return i, j
+            end
+        end
+    end
+end
+
+local function index_to_point(data, index)
+    return math.floor(index / data._private.row_count) + 1,
+        index % data._private.column_count + 1
+end
+
+--- Add some widgets to the given grid layout
+-- @param ... Widgets that should be added (must at least be one)
+function grid:add(...)
+
+    local args = { n=select('#', ...), ... }
+    assert(args.n > 0, "need at least one widget to add")
+
+    local row, column
+
+    for i=1, args.n do
+        local w = args[i]
+
+        row, column = get_next_empty(self, row, column)
+
+        base.check_widget(w)
+
+        self._private.rows[row][column] = w
+        table.insert(self._private.children, w)
+    end
+
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Set a widget at a specific position.
+-- @tparam number row The row
+-- @tparam number column The column
+-- @param widget The widget
+function grid:set_cell_widget(row, column, widget)
+    if row < self._private.row_count then
+        grid:set_row_count(row)
+    end
+
+    if column < self._private.column_count then
+        grid:set_column_count(column)
+    end
+
+    self._private.rows[row] = self._private.rows[row] or {}
+
+    self:emit_signal("widget::layout_changed")
+end
+
+--- The number of columns.
+-- @property column_count
+-- @param number
+
+--- The number of rows.
+-- Note that this is automatically incremented when more elements
+-- are added to the layout.
+-- @property row_count
+-- @param number
+
+function grid:set_row_count(count)
+    self._private.row_count = count
+    self:emit_signal("widget::layout_changed")
+end
+
+function grid:set_column_count(count)
+    self._private.column_count = count
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Re-pack all the widget according to the current `row_count` and `column_count`.
+function grid:reflow()
+    local children = self:get_children()
+    self:reset()
+    if #children > 0 then
+        self:add(unpack(children))
+    end
```

Comment:
How about implementing this function as `self:set_children(self:get_children())`? Seems simpler to me.